log4j-api/src/main/java/org/apache/logging/log4j/spi/Provider.java

```diff
@@ -97,127 +217,266 @@ public String getVersions() {
+        return new NoOpThreadContextMap();
+    }
+
+    // Used for testing
+    void resetThreadContextMap() {
+        threadContextMapLazy.set(null);
+    }
+
+    /**
+     * @return A lazily initialized thread context map.
+     * @since 2.24.0
+     */
+    public final ThreadContextMap getThreadContextMapInstance() {
+        return threadContextMapLazy.get();
+    }
+
     /**
      * Gets the URL containing this Provider's Log4j details.
      *
-     * @return the URL corresponding to the Provider {@code META-INF/log4j-provider.properties} file
+     * @return the URL corresponding to the Provider {@code META-INF/log4j-provider.properties} file or {@code null}
+     * for a provider class.
+     * @deprecated since 2.24.0, without replacement.
      */
+    @Deprecated
     public URL getUrl() {
         return url;
     }
 
     @Override
     public String toString() {
-        final StringBuilder result = new StringBuilder("Provider[");
+        final StringBuilder result =
+                new StringBuilder("Provider '").append(getClass().getName()).append("'");
         if (!DEFAULT_PRIORITY.equals(priority)) {
-            result.append("priority=").append(priority).append(", ");
+            result.append("\n\tpriority             = ").append(priority);
         }
+        final String threadContextMap = getThreadContextMap();
         if (threadContextMap != null) {
-            result.append("threadContextMap=").append(threadContextMap).append(", ");
-        } else if (threadContextMapClass != null) {
-            result.append("threadContextMapClass=").append(threadContextMapClass.getName());
+            result.append("\n\tthreadContextMap     = ").append(threadContextMap);
         }
-        if (className != null) {
-            result.append("className=").append(className).append(", ");
-        } else if (loggerContextFactoryClass != null) {
-            result.append("class=").append(loggerContextFactoryClass.getName());
+        final String loggerContextFactory = getClassName();
+        if (loggerContextFactory != null) {
+            result.append("\n\tloggerContextFactory = ").append(loggerContextFactory);
         }
         if (url != null) {
-            result.append("url=").append(url);
+            result.append("\n\turl                  = ").append(url);
         }
-        final ClassLoader loader;
-        if (classLoader == null || (loader = classLoader.get()) == null) {
-            result.append(", classLoader=null(not reachable)");
-        } else {
-            result.append(", classLoader=").append(loader);
+        if (Provider.class.equals(getClass())) {
+            final ClassLoader loader = classLoader.get();
+            if (loader == null) {
+                result.append("\n\tclassLoader          = null or not reachable");
+            } else {
+                result.append("\n\tclassLoader          = ").append(loader);
```

Comment:
Please don't do this type of indentation. It is a matter of time somebody will screw up the alignment. I prefer the earlier version.